File: miscsrc/parse_mvt_sequences.py
#!/usr/bin/env python3

import argparse
import sys
import mapbox_vector_tile
import json
import math
import re
from datetime import datetime

def parse_date(datestr):
    """Convert YYYYMMDD to YYYY-MM-DD if format matches."""
    if not datestr:
        return None
    m = re.match(r"^(\d{4})(\d{2})(\d{2})$", datestr)
    if not m:
        print("Invalid date format: {} (expected YYYYMMDD)".format(datestr), file=sys.stderr)
        sys.exit(1)
    return "{}-{}-{}".format(m.group(1), m.group(2), m.group(3))

def tile_coord_to_latlon(tile_x, tile_y, zoom, geom_x, geom_y):
    tile_size = 4096
    n = 2 ** zoom

    # Flip Y coordinate within tile to correct mirrored coordinates
    flipped_y = tile_size - geom_y

    pixel_x = tile_x * tile_size + geom_x
    pixel_y = tile_y * tile_size + flipped_y

    lon_deg = (pixel_x / (tile_size * n)) * 360.0 - 180.0
    merc_y = (pixel_y / (tile_size * n))
    lat_rad = math.atan(math.sinh(math.pi * (1 - 2 * merc_y)))
    lat_deg = math.degrees(lat_rad)

    return lat_deg, lon_deg

def parse_filename(filename):
    match = re.search(r'tile_(\d+)_(\d+)_(\d+)\.mvt$', filename)
    if not match:
        print("Error: Filename must be in format tile_<zoom>_<tile_x>_<tile_y>.mvt")
        sys.exit(1)
    zoom = int(match.group(1))
    tile_x = int(match.group(2))
    tile_y = int(match.group(3))
    return zoom, tile_x, tile_y

def flatten_coordinates(coords):
    if isinstance(coords[0], (int, float)):
        return [coords]
    else:
        result = []
        for c in coords:
            result.extend(flatten_coordinates(c))
        return result

def format_timestamp_ms(ts):
    try:
        # Convert milliseconds to integer seconds
        dt = datetime.utcfromtimestamp(round(ts / 1000.0))
        return dt.isoformat()
    except (ValueError, OSError):
        return ""

def main():
    parser = argparse.ArgumentParser(
        description="Process MVT tile files with optional date filtering."
    )
    parser.add_argument(
        "tile_file",
        help="Path to the tile file (.mvt)",
    )
    parser.add_argument(
        "date_from",
        nargs="?",
        help="Start date (format: YYYYMMDD)",
    )
    parser.add_argument(
        "date_to",
        nargs="?",
        help="End date (format: YYYYMMDD)",
    )
    parser.add_argument(
        "--debug",
        action="store_true",
        help="Enable debug mode (prints extra information)",
    )

    args = parser.parse_args()

    filter_date_from = parse_date(args.date_from)
    filter_date_to   = parse_date(args.date_to)

    if args.debug:
        print("date_from: {}".format(json.dumps(filter_date_from)), file=sys.stderr)
        print("date_to:   {}".format(json.dumps(filter_date_to)),   file=sys.stderr)

    tile_file = args.tile_file
    zoom, tile_x, tile_y = parse_filename(tile_file)

    with open(tile_file, "rb") as f:
        tile_bytes = f.read()

    tile = mapbox_vector_tile.decode(tile_bytes)
    sequence_layer = tile.get("sequence")
    if not sequence_layer:
        print("No 'sequence' layer found in tile.")
        sys.exit(0)

    features = sequence_layer.get("features", [])
    sequences = []

    for feat in features:
        if args.debug:
            print(json.dumps(feat, indent=2), file=sys.stderr)

        if not isinstance(feat, dict):
            continue

        props = feat.get("properties", {})
        geom = feat.get("geometry", {})
        coords = geom.get("coordinates", [])
        seq_id = props.get("id")

        if not coords or not seq_id:
            if args.debug:
                print("skipping feature, no coords/seq id...", file=sys.stderr)
            continue

        start_id = props.get("image_id") or props.get("start_id") or seq_id
        creator_id = props.get("creator_id")
        creator = str(creator_id) if creator_id is not None else ""
        make = props.get("make") or ""
        start_captured_at = props.get("captured_at") or props.get("start_captured_at") or ""
        end_captured_at = props.get("end_captured_at") or ""

        formatted_start = format_timestamp_ms(start_captured_at) if start_captured_at else ""

        # Filter sequences by hardcoded start date
        if formatted_start:
            if filter_date_from and formatted_start < filter_date_from:
                if args.debug:
                    print("skipping feature, before date_from...", file=sys.stderr)
                continue  # skip sequences before start date
            if filter_date_to and formatted_start > filter_date_to:
                if args.debug:
                    print("skipping feature, after date_to...", file=sys.stderr)
                continue  # skip sequences after start date
        else:
            if args.debug:
                print("skipping feature, no start date...", file=sys.stderr)
            continue  # skip if no start date

        flat_coords = flatten_coordinates(coords)
        print(json.dumps(flat_coords))
        latlon_points = []
        for coord in flat_coords:
            geom_x, geom_y = coord[0], coord[1]
            lat, lon = tile_coord_to_latlon(tile_x, tile_y, zoom, geom_x, geom_y)
            latlon_points.append([lat, lon])

        date_from = formatted_start[:10] if formatted_start else ""
        date_to = date_from
        formatted_end = format_timestamp_ms(end_captured_at) if end_captured_at else ""

        url = (
            "https://www.mapillary.com/app/user/{creator}"
            "?pKey={start_id}&focus=photo&dateFrom={date_from}&dateTo={date_to}"
            "&z=15&lat={lat}&lng={lon}"
        ).format(
            creator=creator,
            start_id=start_id,
            date_from=date_from,
            date_to=date_to,
            lat=latlon_points[0][0] if latlon_points else 0,
            lon=latlon_points[0][1] if latlon_points else 0,
        )

        sequences.append({
            "url": url,
            "start_captured_at": formatted_start,
            "end_captured_at": formatted_end,
            "creator": creator,
            "make": make,
            "start_id": start_id,
            "sequence": seq_id,
            "coordinates": latlon_points,
        })

    print(json.dumps(sequences, indent=2))

if __name__ == "__main__":
    main()

File: miscsrc/parse_mvt_sequences.pl
#!/usr/bin/env perl
# -*- perl -*-

#
# Author: Slaven Rezic
#
# Copyright (C) 2025 Slaven Rezic. All rights reserved.
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
#
# WWW:  https://github.com/eserte/bbbike
#

use strict;
use warnings;
use FindBin;
use v5.10.0; # defined-or

use JSON::PP;
use Math::Trig 'pi', 'sinh', 'atan';
use POSIX 'strftime';

my $miscdir = "$FindBin::RealBin/../misc";

# Helper: parse YYYYMMDD date format to comparable string
sub parse_date {
  my $datestr = shift;
  return unless defined $datestr;
  if ($datestr =~ /^(\d{4})(\d{2})(\d{2})$/) {
    return "$1-$2-$3";
  }
  die "Invalid date format, must be YYYYMMDD";
}

# Helper: extract zoom, tile_x, tile_y from filename
sub parse_filename {
  my $filename = shift;
  if ($filename =~ /tile_(\d+)_(\d+)_(\d+)\.mvt$/) {
    return ($1, $2, $3);
  }
  die "Filename must be tile_<zoom>_<x>_<y>.mvt";
}

# Helper: convert tile and geometry coords to lat/lon
sub tile_coord_to_latlon {
  my ($tile_x, $tile_y, $zoom, $geom_x, $geom_y) = @_;
  my $tile_size = 4096;
  my $n = 2 ** $zoom;

  my $flipped_y = $tile_size - $geom_y;
  my $pixel_x = $tile_x * $tile_size + $geom_x;
  my $pixel_y = $tile_y * $tile_size + $flipped_y;

  my $lon_deg = ($pixel_x / ($tile_size * $n)) * 360.0 - 180.0;
  my $merc_y = ($pixel_y / ($tile_size * $n));
  my $lat_rad = atan(sinh(pi * (1 - 2 * $merc_y)));
  my $lat_deg = $lat_rad * 180 / pi;

  return ($lat_deg, $lon_deg);
}

# Helper: recursively flatten nested geometry coordinates
sub flatten_coordinates {
  my $coords = shift;
  if (ref $coords->[0] ne 'ARRAY') { return [$coords]; }
  my @flat;
  for my $c (@$coords) { push @flat, @{flatten_coordinates($c)}; }
  return \@flat;
}

# Helper: convert captured_at timestamps from milliseconds
sub format_timestamp_ms {
  my $ts = shift;
  return "" if !defined $ts or $ts eq '';
  my $sec = int($ts / 1000);
  # Convert to ISO8601 UTC timestamp
  return strftime("%Y-%m-%dT%H:%M:%SZ", gmtime($sec));
}

use constant {
  CMD_MOVE_TO  => 1,
  CMD_LINE_TO  => 2,
  CMD_CLOSE_PATH => 7,
};

# ZigZag decode an integer
sub zigzag_decode {
  my $n = shift;
  use integer;
  return ($n >> 1) ^ ( -($n & 1) );
}

#sub zigzag_decode {
#  my $n = shift;
#use Devel::Peek; Dump $n;#XXX
#my $r = ($n >> 1) ^ ( -($n & 1) );
#warn $r;
#  # Perl integers are platform dependent, ensure proper signed conversion:
#  return ($n >> 1) ^ ( -($n & 1) );
#}

#sub zigzag_decode {
#    my ($z) = @_;
#    my $sign = $z & 1;             # 0 for positive, 1 for negative
#    my $value = $z >> 1;           # logical shift right
#    return $sign ? -($value + 1) : $value;
#}



# Decode geometry commands from array of integers
# Returns arrayref of parts, each part is arrayref of [x,y] points
sub decode_geometry {
  my($geom, $extent) = @_;
  my @geometry = @$geom;  # array of integers from protobuf
  my $i = 0;
  my $dx = 0;
  my $dy = 0;
  my @parts;

  while ($i < @geometry) {
    my $command_integer = $geometry[$i++];
    my $cmd = $command_integer & 0x7;
    my $count = $command_integer >> 3;
    die "Unknown command $cmd" unless $cmd == CMD_MOVE_TO || $cmd == CMD_LINE_TO || $cmd == CMD_CLOSE_PATH;

    if ($cmd == CMD_CLOSE_PATH) {
      # ClosePath has no params and count is usually 1
      # It means close the current ring (connect last point to first)
      # We don't need params, just mark last part closed
      # Optionally you may close the polygon ring explicitly in your usage
      next;
    }

    my @points;
    for (1..$count) {
      my $x = zigzag_decode($geometry[$i++]);
      my $y = zigzag_decode($geometry[$i++]);
      $x += $dx;
      $y += $dy;

      $dx = $x;
      $dy = $y;
      
      # if not y_coord_down:
      $y = $extent - $y;

      push @points, [$x, $y];
    }

    if ($cmd == CMD_MOVE_TO) {
      # MoveTo starts a new part
      push @parts, \@points;
    }
    elsif ($cmd == CMD_LINE_TO) {
      # LineTo appends points to the current part
      if (@parts) {
        push @{$parts[-1]}, @points;
      } else {
        # Usually should not happen, but in case no prior MoveTo, treat as new part
        push @parts, \@points;
      }
    }
  }

  return \@parts;
}

# Main processing logic
my ($zoom, $tile_x, $tile_y) = parse_filename($ARGV[0]);
my $date_from = $ARGV[1] ? parse_date($ARGV[1]) : undef;
my $date_to = $ARGV[2] ? parse_date($ARGV[2]) : undef;

open my $fh, "<:raw", $ARGV[0] or die $!;
my $data = do { local $/; <$fh> };

my $proto = do { open my $fh, "$miscdir/vector_tile.proto" or die "Can't load $miscdir/vector_tile.proto: $!"; local $/; <$fh> };

my $tile;
if (eval { require Google::ProtocolBuffers::Dynamic; 1 }) {
    #warn "Use Google::ProtocolBuffers::Dynamic.\n";
    my $pb = Google::ProtocolBuffers::Dynamic->new($FindBin::RealBin);
    $pb->load_string("vector_tile.proto", <<EOF . $proto);
syntax = "proto2";
EOF
    $pb->map({ package => 'vector_tile', prefix => 'VectorTile' });

    $tile = VectorTile::Tile->decode($data);
} else {
    require Google::ProtocolBuffers;
    #warn "Use Google::ProtocolBuffers.\n";
    Google::ProtocolBuffers->parse($proto);

    $tile = VectorTile::Tile->decode($data);
}
#require Data::Dumper; print STDERR "Line " . __LINE__ . ", File: " . __FILE__ . "\n" . Data::Dumper->new([$tile],[qw()])->Indent(1)->Useqq(1)->Sortkeys(1)->Terse(1)->Dump; # XXX

my ($sequence_layer) = grep { $_->{name} eq 'sequence' } @{$tile->{layers}};
die "No 'sequence' layer found in tile." unless $sequence_layer;

my @sequences;

my($keys, $values) = ($sequence_layer->{keys}, $sequence_layer->{values});
my $extent = $sequence_layer->{extent} || die "extent missing in layer";

for my $feat (@{$sequence_layer->{features}}) {

  my $tags = $feat->{tags};
  my $props = {};
  for (my $i = 0; $i < @$tags; $i += 2) {
    my $key_idx = $tags->[$i];
    my $val_idx = $tags->[$i+1];

    my $key = $keys->[$key_idx];
    my $val_entry = $values->[$val_idx];

    # Values are typed, e.g., $val_entry->{string_value}, {int_value}, etc.
    # Extract the actual value depending on which field is set
    my $value;
    if (exists $val_entry->{string_value}) {
	$value = $val_entry->{string_value};
    }
    elsif (exists $val_entry->{int_value}) {
	$value = $val_entry->{int_value};
    }
    elsif (exists $val_entry->{double_value}) {
	$value = $val_entry->{double_value};
    }
    elsif (exists $val_entry->{bool_value}) {
	$value = $val_entry->{bool_value};
    }
    # ... handle other value types similarly

    $props->{$key} = $value;
  }

  my $seq_id = $props->{id} or next;

  my $geom_data = $feat->{geometry};
  my $coords = decode_geometry($geom_data, $extent);
  $coords or next;

  # Flatten and convert coordinates
  my $flat_coords = flatten_coordinates($coords);
#require Data::Dumper; print STDERR "Line " . __LINE__ . ", File: " . __FILE__ . "\n" . Data::Dumper->new([$flat_coords],[qw()])->Indent(1)->Useqq(1)->Sortkeys(1)->Terse(1)->Dump; # XXX
  my @latlon;
  for my $coord (@$flat_coords) {
    my ($lat, $lon) = tile_coord_to_latlon($tile_x, $tile_y, $zoom, $coord->[0], $coord->[1]);
    push @latlon, [$lat, $lon];
  }

  my $start_captured_at = format_timestamp_ms($props->{captured_at} // $props->{start_captured_at});
  next unless $start_captured_at; # Filter no start date

  if ($date_from && $start_captured_at lt $date_from) { next; }
  if ($date_to && $start_captured_at gt $date_to) { next; }

  my $start_id = $props->{image_id} // $props->{start_id} // $seq_id;
  my $creator = defined $props->{creator_id} ? $props->{creator_id} : "";
  my $make = $props->{make} // "";
  my $end_captured_at = format_timestamp_ms($props->{end_captured_at} // "");

  my $url = sprintf("https://www.mapillary.com/app/user/%s?pKey=%s&focus=photo&dateFrom=%s&dateTo=%s&z=15&lat=%f&lng=%f",
                    $creator, $start_id, substr($start_captured_at,0,10), substr($start_captured_at,0,10),
                    $latlon[0][0], $latlon[0][1]);

  push @sequences, {
    url => $url,
    start_captured_at => $start_captured_at,
    end_captured_at => $end_captured_at,
    creator => $creator,
    make => $make,
    start_id => $start_id,
    sequence => $seq_id,
    coordinates => \@latlon,
  };
}

print encode_json(\@sequences);


__END__
